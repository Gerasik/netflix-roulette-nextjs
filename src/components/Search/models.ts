import * as SearchModels from 'containers/Search/models';
import * as BodyModels from 'containers/Body/models';

interface SearchData {
  searchString: string;
  searchBy: string;
  sortBy: string;
}

export interface SearchProps {
  searchData: SearchData;
  changeSearchString: SearchModels.ActionCreator.ChangeSearchString;
  changeSearchBy: SearchModels.ActionCreator.ChangeSearchBy;
  setStartData: BodyModels.ActionCreator.SetStartData;
}
