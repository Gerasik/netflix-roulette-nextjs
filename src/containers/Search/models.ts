import Immutable from 'immutable';

import * as Models from 'models';

export type SearchState = {
  searchString: string;
  searchBy: string;
  sortBy: string;
};

export type State = Immutable.Record<SearchState>;

export namespace Payload {
  export type ChangeSearchString = string;
  export type ChangeSearchBy = string;
  export type ChangeSortBy = string;
}

export namespace Action {
  export type ChangeSearchString = Models.IAction<Payload.ChangeSearchString>;
  export type ChangeSearchBy = Models.IAction<Payload.ChangeSearchBy>;
  export type ChangeSortBy = Models.IAction<Payload.ChangeSortBy>;
}
export namespace ActionCreator {
  export type ChangeSearchString = (
    payload: Payload.ChangeSearchString
  ) => Action.ChangeSearchString;
  export type ChangeSearchBy = (payload: Payload.ChangeSearchBy) => Action.ChangeSearchBy;
  export type ChangeSortBy = (payload: Payload.ChangeSortBy) => Action.ChangeSortBy;
}
