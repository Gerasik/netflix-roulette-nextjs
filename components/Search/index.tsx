import React, { FunctionComponent } from 'react';
import classNames from 'classnames';

import { Action } from 'containers/Search/models';
import styles from './style.module.scss';
import { SearchProps } from './models';

const Search: FunctionComponent<SearchProps> = ({
  searchData,
  changeSearchString,
  changeSearchBy,
  setData,
}) => {
  const { searchString, searchBy, sortBy } = searchData;

  function handleClickSearch(): void {
    fetch(
      `https://reactjs-cdp.herokuapp.com/movies?sortBy=${sortBy}&sortOrder=asc&search=${searchString}&searchBy=${searchBy}&limit=6`
    )
      .then(response => response.json())
      .then(response => {
        setData(response);
      });
  }

  return (
    <div className={styles.root}>
      <p className={styles.label}>find your movie</p>
      <div className={styles['search-form']}>
        <input
          type="text"
          placeholder="Search"
          className={styles['search-text']}
          onChange={(event): Action.ChangeSearchString => changeSearchString(event.target.value)}
          value={searchString}
        />
        <button type="button" className={styles['search-button']} onClick={handleClickSearch}>
          search
        </button>
      </div>
      <div className={styles['search-by']}>
        <span>search by</span>
        <button
          type="button"
          className={classNames(styles['by-title'], { [styles.active]: searchBy === 'title' })}
          onClick={(): Action.ChangeSearchBy => changeSearchBy('title')}
        >
          title
        </button>
        <button
          type="button"
          className={classNames(styles['by-genre'], { [styles.active]: searchBy === 'genres' })}
          onClick={(): Action.ChangeSearchBy => changeSearchBy('genres')}
        >
          genre
        </button>
      </div>
    </div>
  );
};

export default Search;
