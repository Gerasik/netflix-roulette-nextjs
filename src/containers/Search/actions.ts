import ActionType from './constants';
import { ActionCreator } from './models';

export const changeSearchString: ActionCreator.ChangeSearchString = payload => ({
  type: ActionType.CHANGE_SEARCH_STRING,
  payload,
});

export const changeSearchBy: ActionCreator.ChangeSearchBy = payload => ({
  type: ActionType.CHANGE_SEARCH_BY,
  payload,
});

export const changeSortBy: ActionCreator.ChangeSortBy = payload => ({
  type: ActionType.CHANGE_SORT_BY,
  payload,
});
