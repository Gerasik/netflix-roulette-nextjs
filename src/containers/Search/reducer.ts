import Immutable from 'immutable';
import { REHYDRATE } from 'redux-persist/constants';

import * as Models from 'models';
import { Action, State } from './models';
import ActionType from './constants';

const initialState: State = Immutable.fromJS({
  searchString: '',
  searchBy: 'title',
  sortBy: 'vote_average',
});

const searchReducer = (state = initialState, action: Models.IAction): State => {
  switch (action.type) {
    case ActionType.CHANGE_SEARCH_STRING:
      return changeSearchString(state, action);

    case ActionType.CHANGE_SEARCH_BY:
      return changeSearchBy(state, action);

    case ActionType.CHANGE_SORT_BY:
      return changeSortBy(state, action);

    case REHYDRATE: {
      const incoming = action.payload.searchReducer;
      if (incoming) return incoming;
      return state;
    }

    default:
      return state;
  }
};

export default searchReducer;

export const changeSearchString: Models.Reducer<State, Action.ChangeSearchString> = (
  state,
  action
) => {
  return state.set('searchString', action.payload);
};

export const changeSearchBy: Models.Reducer<State, Action.ChangeSearchBy> = (state, action) => {
  return state.set('searchBy', action.payload);
};

export const changeSortBy: Models.Reducer<State, Action.ChangeSortBy> = (state, action) => {
  return state.set('sortBy', action.payload);
};
