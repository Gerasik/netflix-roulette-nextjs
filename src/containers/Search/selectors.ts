import { createSelector } from 'reselect';

import { State } from './models';

export const searchInfo = (state): State => {
  return state.get('searchReducer') as State;
};

export const searchSortBy = (state): string => {
  return searchInfo(state).get('sortBy');
};

export const searchData = createSelector(searchInfo, searchInfo => ({
  searchString: searchInfo.get('searchString'),
  searchBy: searchInfo.get('searchBy'),
  sortBy: searchInfo.get('sortBy'),
}));
