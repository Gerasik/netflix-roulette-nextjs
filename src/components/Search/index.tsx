import React, { FunctionComponent, useEffect } from 'react';
import classNames from 'classnames';
import { debounce } from 'lodash';
import { useRouter } from 'next/router';

import { Action } from 'containers/Search/models';
import styles from './style.module.scss';
import { SearchProps } from './models';

const Search: FunctionComponent<SearchProps> = ({
  searchData,
  changeSearchString,
  changeSearchBy,
  setStartData,
}) => {
  const { searchBy, searchString } = searchData;
  const router = useRouter();
  const { query } = router.query;

  useEffect(() => {
    changeSearchString(query as string);
    setStartData();
  }, [query]);

  function handleClickSearch(): void {
    setStartData();
  }

  function handleChangeSearch(event): void {
    const searchString = event.target.value;
    const href =
      window.location.href.slice(-6) === 'search' ? `search/${searchString}` : searchString;

    window.history.pushState(null, '', href || '/search');
    setStartData();
  }

  function debounceEvent(...arg) {
    const fn = arg[0];
    const debounceEvent = debounce(fn, 500);
    return e => {
      changeSearchString(e.target.value);
      e.persist();
      return debounceEvent(e);
    };
  }

  return (
    <div className={styles.root}>
      <p className={styles.label}>find your movie</p>
      <div className={styles['search-form']}>
        <input
          value={searchString}
          type="text"
          placeholder="Search"
          className={styles['search-text']}
          onChange={debounceEvent(handleChangeSearch)}
        />
        <button type="button" className={styles['search-button']} onClick={handleClickSearch}>
          search
        </button>
      </div>
      <div className={styles['search-by']}>
        <span>search by</span>
        <button
          type="button"
          className={classNames(styles['by-title'], { [styles.active]: searchBy === 'title' })}
          onClick={(): Action.ChangeSearchBy => changeSearchBy('title')}
        >
          title
        </button>
        <button
          type="button"
          className={classNames(styles['by-genre'], { [styles.active]: searchBy === 'genres' })}
          onClick={(): Action.ChangeSearchBy => changeSearchBy('genres')}
        >
          genre
        </button>
      </div>
    </div>
  );
};

export default Search;
