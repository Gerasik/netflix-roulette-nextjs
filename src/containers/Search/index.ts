import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';

import Search from 'components/Search';
import { setStartData } from 'containers/Body/actions';
import { changeSearchString, changeSearchBy } from './actions';
import { searchData } from './selectors';

const mapStateToProps = createStructuredSelector({
  searchData,
});

const mapDispatchToProps = {
  changeSearchString,
  changeSearchBy,
  setStartData,
};

export default connect(mapStateToProps, mapDispatchToProps)(Search);
