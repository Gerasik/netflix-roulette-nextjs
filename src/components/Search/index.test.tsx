import React from 'react';
import { mount } from 'enzyme';
import { createBrowserHistory } from 'history';
import { Router } from 'react-router-dom';

import { testSearchReducerData, testMovieResponse } from 'mock';
import Search from '.';

const history = createBrowserHistory();
const searchComponent = (
  <Router history={history}>
    <Search
      searchData={testSearchReducerData}
      changeSearchString={() => ({ type: 'string', payload: 'testMovieResponse' })}
      changeSearchBy={() => ({ type: 'string', payload: 'testMovieResponse' })}
      setStartData={() => ({ type: 'string', payload: testMovieResponse })}
    />
  </Router>
);

it('render correctly date component', () => {
  const wrapper = mount(searchComponent);
  wrapper.find('.search-button').simulate('click');
  wrapper.find('.by-title').simulate('click');
  wrapper.find('.by-genre').simulate('click');
  wrapper.find('.search-text').simulate('change', { target: { value: 'string' } });
  expect(wrapper).toMatchSnapshot();
});

describe('<Search />', () => {
  it('renders an `footer`', () => {
    const wrapper = mount(searchComponent);
    expect(wrapper.find('input').prop('placeholder')).toBe('Search');
  });
});
