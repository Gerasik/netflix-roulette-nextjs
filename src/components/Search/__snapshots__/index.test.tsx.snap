// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly date component 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Search
    changeSearchBy={[Function]}
    changeSearchString={[Function]}
    searchData={
      Object {
        "searchBy": "genres",
        "searchString": "",
        "sortBy": "vote_average",
      }
    }
    setStartData={[Function]}
  >
    <p
      className="label"
    >
      find your movie
    </p>
    <div
      className="search-form"
    >
      <input
        className="search-text"
        onChange={[Function]}
        placeholder="Search"
        type="text"
      />
      <button
        className="search-button"
        onClick={[Function]}
        type="button"
      >
        search
      </button>
    </div>
    <div
      className="search-by"
    >
      <span>
        search by
      </span>
      <button
        className="by-title"
        onClick={[Function]}
        type="button"
      >
        title
      </button>
      <button
        className="by-genre active"
        onClick={[Function]}
        type="button"
      >
        genre
      </button>
    </div>
  </Search>
</Router>
`;
